base/dcp_common.go

```diff
@@ -613,32 +614,70 @@ func alternateAddressShims(bucketSpecTLS bool) (
 			return nil, err
 		}
 
+		// Fetch any alternate external addresses/ports and store them in the externalAlternateAddresses map
 		pool, err := client.GetPool(poolName)
 		if err != nil {
 			return nil, err
 		}
 
+		poolServices, err := client.GetPoolServices(poolName)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(pool.Nodes) != len(poolServices.NodesExt) {
+			return nil, fmt.Errorf("length of pool nodes list is not equal to poolServices nodes list: poolNodes: %v, poolServices.NodesExt: %v", pool.Nodes, poolServices.NodesExt)
+		}
+
+		connSpecAddressesHostMap := make(map[string]struct{}, len(connSpecAddresses))
+		for _, connSpecAddress := range connSpecAddresses {
+			connSpecAddressesHostMap[connSpecAddress.Host] = struct{}{}
+		}
+
 		// Recreate the map to forget about previous clustermap information.
-		externalAlternateAddresses = make(map[string]string, len(pool.Nodes))
-		for _, node := range pool.Nodes {
+		externalAlternateAddresses = make(map[string]string, len(poolServices.NodesExt))
+		for i, node := range poolServices.NodesExt {
+			Tracef(KeyDCP, "ps.NodesExt[%d]: %v", i, MD(node))
 			if external, ok := node.AlternateNames["external"]; ok && external.Hostname != "" {
-				nodeHostname, _, err := SplitHostPort(node.Hostname)
+				// if hostname != match, continue/skip
+				if _, ok := connSpecAddressesHostMap[external.Hostname]; !ok {
+					Tracef(KeyDCP, "hostname %q not in connSpecAddressesHostMap: %#v", external.Hostname, connSpecAddressesHostMap)
```

Comment:
This looks like that in order to use the external address for a given node, users need to include the external address for *all* Couchbase Server nodes in Sync Gateway's config.  That's not ideal - we'd like to be able to modify the number of nodes in the CBS cluster without requiring an update to the SG config.  

I think it would be preferable to check the set of addresses defined in connSpecAddresses to see whether they are alternate addresses or not.   If they are all alternate, use alternate for all nodes.  If they are all non-alternate, use non-alternate for all nodes.  If they are a mix, use alternate only for those explicitly defined that way in the connStr, and non-alternate for the rest.

Reply:
I did testing around this; both specifying multi-node connStr for internal and external addresses mixed in the SG config, as well single node for both, having only a partial set of CB nodes using an alt address, as well as only a partial set of nodes with a KV service.

I'll do a bit more testing, and re-read the SDK/XDCR/Server specs so I can answer these questions definitively tomorrow.